File: domain/src/main/java/cn/byronlab/weather/domain/usecase/SearchCitiesUseCase.java
package cn.byronlab.weather.domain.usecase;

import cn.byronlab.weather.domain.model.City;
import cn.byronlab.weather.domain.repository.CityRepository;
import cn.byronlab.weather.domain.result.DomainResult;

import java.util.Collections;
import java.util.List;

public final class SearchCitiesUseCase {

    private final CityRepository cityRepository;

    public SearchCitiesUseCase(CityRepository cityRepository) {
        this.cityRepository = cityRepository;
    }

    public DomainResult<List<City>> execute(String keyword) {
        String normalizedKeyword = UseCasePreconditions.trim(keyword);
        if (normalizedKeyword.isEmpty()) {
            return DomainResult.success(Collections.emptyList());
        }
        return cityRepository.searchCities(normalizedKeyword);
    }
}

File: domain/src/main/java/cn/byronlab/weather/domain/usecase/SearchCitiesUseCase.kt
package cn.byronlab.weather.domain.usecase

import cn.byronlab.weather.domain.model.City
import cn.byronlab.weather.domain.repository.CityRepository
import cn.byronlab.weather.domain.result.DomainResult
import javax.inject.Inject

class SearchCitiesUseCase @Inject constructor(
    private val cityRepository: CityRepository,
) {

    suspend operator fun invoke(keyword: String?): DomainResult<List<City>> {
        val normalizedKeyword = UseCasePreconditions.trim(keyword)
        if (normalizedKeyword.isEmpty()) {
            return DomainResult.success(emptyList())
        }
        return cityRepository.searchCities(normalizedKeyword)
    }
}
